torch/lib/c10d/ProcessGroupMPI.cpp

```diff
@@ -588,6 +588,167 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::reduce_scatter(
   throw std::runtime_error("ProcessGroupMPI does not support reduce_scatter");
 }
 
+std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::alltoall_base(
+    std::vector<at::Tensor>& outputTensors,
+    std::vector<at::Tensor>& inputTensors,
```

Comment:
let's pass a single tensor here instead to keep API consistent (i.e. similar to allgather_base)?

Reply:
Sure, I am no sure why other APIs take a vector of tensors when single tensor is expected. I thought `allgather_base` would change eventually. But if that's not the case, i will change definition of `alltoall_base` to align with `allgather_base`. Also, would you consider extending `allgather_base` to support gathering uneven sized tensors using `output_spilt_sizes`? We see need for this in distributed DLRM when global batch size is not evenly split across available ranks.

Reply:
this wasn't on our roadmap. AFAIK, some backends (NCCL) expect same size tensors 
https://docs.nvidia.com/deeplearning/sdk/nccl-developer-guide/docs/api/colls.html#ncclallgather

Can you pad tensors for DLRM usecase, would it lead to significant loss of efficiency?

Reply:
> Can you pad tensors for DLRM usecase, would it lead to significant loss of efficiency?

Typically, it happens only for last batch and currently we are simply skipping that batch. I think it is not very critical but just a nice to have. But from API point of view, we should keep the flexibility to support it.

Reply:
@agolynski when tables are equally distributed we should see this behavior in the last batch. But the need for alltoallv arises when we do uneven distribution of embedding tables (different number of tables per GPU or node). Therefore I would prefer, like @ddkalamk suggested, to the keep the API flexible. 

> this wasn't on our roadmap. AFAIK, some backends (NCCL) expect same size tensors
NCCL, unlike MPI or other comms libraries, does not support the *v variation, i.e. different size tensors, yet. The alltoallv primitive is driven by workload requirements and NCCL might eventually support these primitives at some point.